### ISSUE TITLE
Changing Window.MousePosition onMotion calls back to onMotion

### ISSUE THREAD

**Issue by eugeneloza** [eugeneloza/decoherence#160]
`Window.MousePosition := WindowCenter;` Should investigate this issue further. Dobule-calling of `onMotion` shouldn't happen maybe I'm doing something wrong?

**Comment by eugeneloza** [comment]
do a temporary fix by a variable: boolean; for Pre-Alpha.

**Comment by michaliskambi** [comment]
It is indeed undefined whether setting `Window.MousePosition` causes `OnMotion` or not. It may even depend on the backend (WinAPI, GTK...).

It is documented on https://castle-engine.sourceforge.io/apidoc/html/CastleWindow.TCastleWindowCustom.html#MousePosition : "It is undefined whether setting mouse position will generate an OnMotion event (just as if the user moved the mouse)."

I'm not happy about this, but I saw no easy and reliable method to make it better, for now. With most window systems, the event caused by changing MousePosition will happen at some time later (not inside SetMousePosition), but even that is undefined.

Bottom line: unfortunately, if you change Window.MousePosition, you need to make your code "robust", to be prepared for an extra OnMotion event (but you cannot assume that it will always happen!). Inside OnMotion, you simply don't know whether the event was caused by a user, or by setting Window.MousePosition from your code.

This *is* a Castle Game Engine bug, to say it clearly. But it's something I really don't see how to fix. The APIs like WinAPI or GTK or Xlib don't make it easy.

**Comment by eugeneloza** [comment]
Thanks a lot for the information!
Then the fix I've just posted should be considered a final solution to the issue :)

> you need to make your code "robust", to be prepared for an extra OnMotion event

That's what I did here. I just check if the `Window.MousePosition` equals to last-assigned position and if it does I just break `onMotion` sequence.

**Comment by eugeneloza** [comment]
@michaliskambi , I've just checked throughout the CGE code, it looks like the issue may be fixed Engine-side by a similar fix.

I see that `OnMotion` is called here:
https://github.com/castle-engine/castle-engine/blob/master/src/components/castlecontrol.pas#L780
and here:
https://github.com/castle-engine/castle-engine/blob/master/src/ui/castleuicontrols.pas#L2447

So if this solution looks fine (see the commit above) I can make it. It doesn't look complex. Just checking if nothing changed in the `Event` before raising `onMotion`, we'll just need to store the "last" `Event` (just as I'm doing now).

`TCastleWindowCustom.SetMousePosition` is described here:
https://github.com/castle-engine/castle-engine/blob/master/src/window/castlewindow_lcl.inc#L581
here:
https://github.com/castle-engine/castle-engine/blob/master/src/window/castlewindow_library.inc#L163
here:
https://github.com/castle-engine/castle-engine/blob/master/src/window/windows/castlewindow_winapi.inc#L989
here:
https://github.com/castle-engine/castle-engine/blob/master/src/window/gtk/castlewindow_gtk.inc#L783
and here:
https://github.com/castle-engine/castle-engine/blob/master/src/window/unix/castlewindow_xlib.inc#L371
(I don't think I've missed anything).

**Comment by michaliskambi** [comment]
Hm, it feels vaguely "dirty" for me, comparing previous and last value. But... well, the current undefined behavior is not a nice situation either. OK, let's do it.

It should not only secure the OnMotion calls, it should secure all the TUIControl.Motion calls.  So it should be done before calling `Container.EventMotion` in

https://github.com/castle-engine/castle-engine/blob/master/src/components/castlecontrol.pas#L1134

or 

https://github.com/castle-engine/castle-engine/blob/master/src/window/castlewindow.pas#L3518

If the new position is the same as old, then you don't need to call `Container.EventMotion`, and of course you don't need to update `FMousePosition`. 

In case of castlewindow.pas, do it only when Event.FingerIndex = 0, to not mess with multi-touch on Android. 

Yes, I think you have found all the SetMousePosition occurrences.

Thank you!

**Comment by eugeneloza** [comment]
Ok. I'll try to make a commit today evening.

**Comment by eugeneloza** [comment]
Hmm... I'm thinking of an alternative (a more universal) way to:

https://github.com/castle-engine/castle-engine/blob/master/src/ui/castleuicontrols.pas#L2413

E.g. we'll have a strict private `FLastMousePosition` in `TUIControl` from which `TContainer` and `TWindowContainer` descend.
It would just stop the `EventMotion` sequence from running and will halt all `onMotion` events. Thou it will still let the `TCastleWindowCustom` and `TCastleControlCustom` resetting it's current mouse position, however, maybe it won't do anything bad? This solution looks cleaner to me, what do you think?
Also it's possible to make `EventMotion` into a function to prevent further assigning `FMousePosition`, calling `UpdateShiftState` and `AggressiveUpdate`.

Alternative way is to use `FLastMousePosition` in both `TCastleWindowCustom` and `TCastleControlCustom`, just it'll be two of them this way and two identical checks in the code, however, it looks like a much better alternative to remaking `EventMotion` into a function.

**Comment by eugeneloza** [comment]
(reopening the issue for now)

**Comment by michaliskambi** [comment]
> E.g. we'll have a strict private FLastMousePosition in TUIControl from which TContainer and TWindowContainer descend.

They descend from TGLContainer, not from TUIControl. TGLContainer descends from TUIContainer, the non-OpenGL logic should go to TUIContainer.

Do you mean you would like to remove existing `FMousePosition` from TCastleWindowCustom and TCastleControlCustom, and replace it with `FLastMousePosition` in TUIContainer? But then, you would need to also 

1. change TCastleWindowCustom.DoMouseDown and TCastleWindowCustom.DoMouseUp, they update FMousePosition too, since window system may not send the OnMotion events with enough "granularity" (which makes sense, e.g. when dragging with finger/mouse you would be flooded with OnMotion events if you received OnMotion for every little mouse/finger move). But you want to have the latest mouse position when the click occurred, so window systems sends it in every MouseDown too.

    Hm, but one can argue that DoMouseDown should then call DoMotion at the beginning...

2. Introduce GetMousePosition in TUIContainer, and make the public MousePosition property in TCastleWindowCustom and TCastleControlCustom use it.

Bottom line: It's probably possible to do it (move more logic to TUIContainer), but as you can see one change pulls another change as a dependency (you change FMousePosition, then you also must change DoMouseDown/Up and MousePosition public property)...  So I would advice to leave it, and do the task in a simple way now (just check `TVector2.PerfectlyEqual(Event.Position, FMousePosition)` in two places, once in CastleWindow and once in CastleControl).

We can do a restructuring later, as a separate pull request. You will get lost if you start to refactor the CastleWindow, there's a lot of code there :)

Note that UpdateShiftState and AggressiveUpdate is only done by TCastleControlCustom (TCastleWindow doesn't need this mechanism).

**Comment by eugeneloza** [comment]
Ok. Got it.

> Do you mean you would like to remove existing `FMousePosition`

No, I've meant to add an additional `FLastMousePosition` (used solely to fix this issue). But as I see now, `FMousePosition` does absolutely the same thing.

**Comment by eugeneloza** [comment]
hmmmm... I've made the patch, but it doesn't seem to solve the problem. Now I understand why.
````
FMousePosition is (0,0)
Mouse moved by (dx,dy) -> FMousePosition<>Event.Position
OnMotion: reset MousePosition to (0,0)
Mouse moved by (-dx,-dy) -> again FMousePosition<>Event.Position
call back to OnMotion.
````

I.e. we need to check for not `TVector2.PerfectlyEqual(Event.Position, FMousePosition)` but `TVector2.PerfectlyEqual(Event.Position, **FRequestedMousePosition**)` I'll see what I can do here.

**Comment by eugeneloza** [comment]
Very robust check in `castlecontrol.pas` and `castleuicontrols.pas` also didn't work... hmm...
```
  if Assigned(OnMotion) and (OnMotionClear) then begin
    OnMotionClear := false;
    OnMotion(Self, Event);
    OnMotionClear := true;
  end;
```

**Comment by michaliskambi** [comment]
> TVector2.PerfectlyEqual(Event.Position, FRequestedMousePosition) I'll see what I can do here.

Yes, my mistake, sorry about that! My initial idea for how this should look like was right :) 

- Every SetMousePosition call should set something like `FRequestedMousePositionValid := true` and `FRequestedMousePosition := requested vector`. You will need to repeat this in 5 code places that you already found, but it's rather trivial :)

- then before calling EventMotion we check whether `(not FRequestedMousePositionValid) or (not FRequestedMousePosition perfectly equals Event.Position)`. 

- In any case, we clear the `FRequestedMousePositionValid` flag.

(Do not try to avoid the boolean flag `FRequestedMousePositionValid` by a special `FRequestedMousePosition`, as every position is valid in principle, even negative ones :)

The inverse check, that I suggested later, `TVector2.PerfectlyEqual(Event.Position, FRequestedMousePosition)`, is indeed not OK.

**Comment by eugeneloza** [comment]
(note) The first patch prevents OnMotion entering an endless loop (as it was before). Thou it's called twice.

I think it's possible to avoid `FRequestedMousePosition perfectly equals Event.Position`, we just need to know if `FMouseWasSetManually: boolean` (set to true in all 6 units) and clear it afterwards.
Thou, at this moment I'm wondering, won't it break the idea of manually controlling the mouse? E.g. making a tutorial for a game, where the cursor is controlled by a program for a few seconds won't generate `OnMotion` and therefore it won't run the code there (e.g. to change the cursor shape if it's over some game element explained in the tutorial).
We just need to eliminate double call-backs of `OnMotion` (and only in case the `MousePosition` was changed INSIDE the `OnMotion`), not to eliminate them at all...
Still thinking on the issue...

**Comment by eugeneloza** [comment]
So the pipeline looks like that.
````
FMousePosition is (0,0)
onMotion: Event.Position is (dx,dy);
(*)
SetMousePosition to (0,0)
--->onMotion: Event.Position is (0,0);
````
Practically the either check above should have fixed this issue clear. Unless in (*) FMousePosition becomes (dx,dy) (and, yes, it looks like it does). I.e. the mouse position goes like that:

`(0,0) -> (dx,dy) -> (0,0)`

failing the "perfectly equals" check.
So, it can't be done in a simple way. The code **must** know if `SetMousePosition` was called from inside `onMotion` (not generating `onMotion` event) or from outside of it (generating `onMotion` event, regardless if anything has changed).

**Comment by michaliskambi** [comment]
> I think it's possible to avoid FRequestedMousePosition perfectly equals Event.Position, we just need to know if FMouseWasSetManually: boolean (set to true in all 6 units) and clear it afterwards.

Not true -- remember that some backends *don't* call OnMotion after SetMousePosition. So blindly using  FMouseWasSetManually would mean that you mistakenly miss some valid EventMotion calls.

> Thou, at this moment I'm wondering, won't it break the idea of manually controlling the mouse?

Right now some backends call OnMotion, and some don't. So this approach is broken anyway, just in other backends.

> Practically the either check above should have fixed this issue clear. 

I don't follow this, probably because I don't know exactly what is the "either check".

Note that the approach with OnMotionClear will only work if the SetMousePosition calls OnMotion immediately inside. Some backends don't do it like this, they place the "mouse motion" event in the queue, and so the OnMotion event is *caused* by SetMousePosition, but happens later.

Although, I'm not really sure whether the full issue is worth fixing... It's probably enough to just fix the possible "recursive calls to OnMotion issue", so that receiving EventMotion from inside SetMousePosition is ignored.

Note that you cannot simply block recursive OnMotion calls. They are valid, e.g. when you do MessageOK from OnMotion. You should only block OnMotion calls that happen inside SetMousePosition (whether this SetMousePosition is inside another OnMotion or not).

**Comment by eugeneloza** [comment]
> 20:28:38> doMotion>doMouseLook: Call-back encountered...
> 20:28:38> 512: 300 (event caused by reset)
> 20:28:38> 513: 300
> 20:28:38> 514: 300
> 20:28:38> doMotion>doMouseLook: Call-back encountered...
> 20:28:38> 512: 300 (event caused by reset)
> 20:28:38> 513: 300
> 20:28:38> doMotion>doMouseLook: Call-back encountered...
> 20:28:38> 512: 300 (event caused by reset)
> 20:28:38> 513: 300
> 20:28:38> doMotion>doMouseLook: Call-back encountered...
> 20:28:38> 512: 300 (event caused by reset)
> 20:28:38> 513: 300
> 20:28:38> 514: 300
> 20:28:38> doMotion>doMouseLook: Call-back encountered...
> 20:28:38> 512: 300 (event caused by reset)
> 20:28:38> 512: 299
> 20:28:38> 513: 299

I've been thinking over this issue, looking at my logs. Maybe, the behaviour we get here IS actually the behaviour we need?
I.e. after the patch above, we will never enter infinite loop, as the Engine cuts off `onMotion` if last position is equal to next one. However, if we reset the position to a DIFFERENT position, then we need to call `onMotion` and it's up for User to keep it safe. But it might be required, as in example I've given above, e.g. this might display the wrong cursor shape after resetting the `MousePosition`.

````
procedure onMotion...
if TVector2.Equals(MousePosition,Vector2(0,0)) then Cursor := mcHand; //this will not happen
if Tutorial = true then MousePosition := Vector2(0,0);
end;
````

And if someone (like me) just needs to reset the cursor, he'll either run the `onMouse` twice or should run a check (like I do) to prevent the second call if the mouse was reset into it's original position.

**Comment by eugeneloza** [comment]
> what is the "either check".

First checking is `TVector2.PerfectlyEquals(FMousePosition,Event.Position)` and second is setting OnMotionClear (now I get why it didn't work, thanks!).

Then it looks like checking `TVector2.PerfectlyEquals(FMousePosition,Event.Position)` prevents the onMotion from entering an infinite loop and makes it safer without actually loosing anything but a few ASM operations.

**Comment by michaliskambi** [comment]
Note that the current behavior is undefined -- people implementing the tutorials like this *should not* trust that OnMotion will happen because of SetMousePosition. They also *should not* assume that it doesn't happen.

Currently, there are 3 situations possible, all of them are really possible (depending on the backend):

1. Inside SetMousePosition, the OnMotion will be called. This is the most dangerous case, as you may end up in an endless loop if your OnMotion unconditionally calls SetMousePosition. Even if your SetMousePosition sets the mouse to the same, exactly equal, position.

2. SetMousePosition places the "make mouse motion" event in the window system event queue, so the OnMotion will be called soon.

3. SetMousePosition does not cause OnMotion at all. Mouse cursor position will change, but we will not be notified about this. If user doesn't touch his mouse, we will not know the new FMousePosition.

It's probably most important to avoid case 1.

Making the rest of the behavior clearly defined (whether we always have case 2, or case 3) is probably not so important.

Hm, maybe it's best to just fix it inside the SetMousePosition? If the requested position is equal to currently known position (in FMousePosition), then ignore the call. Just repeat this test in all SetMousePosition implementations. The only downside is that setting SetMousePosition(0, 0) would be also ignored at the beginning, when FMousePosition is (0, 0) but it's actually "we don't know yet", but that's an edgecase we can probably ignore.

This would not require changing any code around EventMotion at all.

Many possibilities here, and I think that I'm only clouding the issue, not really helping you - sorry about that :) I will leave it up to you...

**Comment by eugeneloza** [comment]
> only clouding the issue, not really helping you

That's very helpful, interesting and a good motivation to learn Engine internals and some interesting specifics :)
I'll try fixing it today a bit later.
